shared/utils/codeql/utils/Locations.qll

```diff
@@ -0,0 +1,100 @@
+
+    /** Gets the 1-based line number (inclusive) where this location starts. */
+    int getStartLine() { Input::locations(this, _, result, _, _, _) }
+
+    /** Gets the 1-based column number (inclusive) where this location starts. */
+    int getStartColumn() { Input::locations(this, _, _, result, _, _) }
+
+    /** Gets the 1-based line number (inclusive) where this location ends. */
+    int getEndLine() { Input::locations(this, _, _, _, result, _) }
+
+    /** Gets the 1-based column number (inclusive) where this location ends. */
+    int getEndColumn() { Input::locations(this, _, _, _, _, result) }
+
+    /** Gets the number of lines covered by this location. */
+    int getNumLines() { result = this.getEndLine() - this.getStartLine() + 1 }
+
+    /** Gets a textual representation of this element. */
+    string toString() {
+      exists(string filepath, int startline, int startcolumn, int endline, int endcolumn |
+        this.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn) and
+        result = filepath + "@" + startline + ":" + startcolumn + ":" + endline + ":" + endcolumn
+      )
+    }
+
+    /**
+     * Holds if this element is at the specified location.
+     * The location spans column `startcolumn` of line `startline` to
+     * column `endcolumn` of line `endline` in file `filepath`.
+     * For more information, see
+     * [LGTM locations](https://codeql.github.com/docs/writing-codeql-queries/providing-locations-in-codeql-queries/).
+     */
+    predicate hasLocationInfo(
+      string filepath, int startline, int startcolumn, int endline, int endcolumn
+    ) {
+      exists(File f |
+        Input::locations(this, f, startline, startcolumn, endline, endcolumn) and
+        filepath = f.getAbsolutePath()
+      )
+    }
+
+    /** Holds if this location starts strictly before the specified location. */
+    pragma[inline]
+    predicate strictlyBefore(Location that) {
+      this.getFile() = that.getFile() and
+      (
+        this.getStartLine() < that.getStartLine()
+        or
+        this.getStartLine() = that.getStartLine() and this.getStartColumn() < that.getStartColumn()
+      )
+    }
+
+    /** Holds if this location ends after location `that`. */
+    pragma[inline]
+    predicate endsAfter(Location that) {
+      this.getFile() = that.getFile() and
+      (
+        this.getEndLine() > that.getEndLine()
+        or
+        this.getEndLine() = that.getEndLine() and this.getEndColumn() > that.getEndColumn()
+      )
+    }
+
+    /**
+     * Holds if this location contains location `that`, meaning that it starts
+     * before and ends after it.
+     */
+    predicate contains(Location that) { this.strictlyBefore(that) and this.endsAfter(that) }
+
+    /** Holds if this location is empty. */
+    predicate isEmpty() { exists(int l, int c | Input::locations(this, _, l, c, l, c - 1)) }
+  }
+
+  /** An entity representing an empty location. */
+  class EmptyLocation extends Location {
```

Comment:
Not all languages may extract empty locations.

Reply:
That's true, not sure if this is a problem though. For those languages the `EmptyLocation` class would simply be empty.

Reply:
> For those languages the EmptyLocation class would simply be empty. 

But that would be confusing.  
I'll move `EmptyLocation` into the language specific bits.